site/content/coursework2.md

```diff
@@ -0,0 +1,424 @@
+# To solve using a W-cycle using 1 iteration of pre-smoothing and 2
+# iterations of post-smoothing to a relative tolerance of 1e-8
+solver.solve(x, b, rtol=1e-8, presmooth=1, postsmooth=2,
+             cycle_type=solver.Type.W)
+```
+
+For a complete example, see the tests in `tests/test_one_dim.py`,
+which implement a one-dimensional example.
+
+The tests themselves do not pass, since the various bits of the `MGSolver`
+class are not completed. You can run the tests with
+[`pytest`](https://pytest.org), using `pytest tests`. After
+successfully implementing the Jacobi iteration and V-cycle, the test
+run looks something like
+```
+$ pytest tests/test_one_dim.py -v
+================================== test session starts ===================================
+platform darwin -- Python 3.8.6, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- pscii/bin/python3
+cachedir: .pytest_cache
+rootdir: XXX, configfile: setup.cfg
+collected 3 items
+
+tests/test_one_dim.py::test_mms_convergence PASSED                                 [ 33%]
+tests/test_one_dim.py::test_two_grid[Jacobi coarse grid] PASSED                    [ 66%]
+tests/test_one_dim.py::test_two_grid[Exact coarse grid] PASSED                     [100%]
+
+=================================== 3 passed in 0.34s ====================================
+```
+
+You can also run the tests in parallel by doing
+```
+$ mpiexec -n 4 tests/test_one_dim.py -v
+```
+
+## Part 1: Reimplementing the beam
+
+In the previous coursework you implemented a 5-point stencil for the Euler-Bernoulli beam
+We will solve the stationary variant of this equation (i.e. no time-component this time).
+In particular, we will solve a three-point bending test. The beam is supported at both ends,
+and at the centre a concentrated load is applied.
+
+$$
+\delta(x-\frac{1}{2}) = -k \partial_{xxxx}u(x)
+$$
+with a small material parameter $k$. Here $\delta$ is a Dirac function, it is one at the centre
+of the beam $x=\frac{1}{2}$ and zero everywhere else.
+
+Boundary conditions remain as before:
+u(0,t)=u(1,t)=0
+$$
+and
+$$
+\partial_{xx}u(0,t) = \partial_{xx}u(1,t) = 0.
+$$
+
+
+To implement this create a class `Bernoulli5pt` that discretises the
+spatial operator using a 2nd order accurate 5-point stencil. Watch out
+for the boundary conditions.
+
+{{< hint info >}}
+Hint: Have a look at `tests/test_one_dim.py`
+{{< /hint >}}
+
+{{< hint info >}}
+You should do your implementation for this part in a file called
+`part1_euler_bernoulli.py` placed in the root of the repository.
+{{< /hint >}}
+
+
+The analytical solution is
+$$w(x) = \begin{cases}
+    \frac{x(4x^2-3)}{48k}, & \mbox{for } 0 < x < \frac{1}{2} \\
+    \frac{(x-1)(1-8x+4x^2)}{48k}, & \mbox{for } \frac{1}{2} < x < 1
+    \end{cases}$$
+but some constant is wrong in the above.
+
+{{< question "Part 1a questions" >}}
+1. How does the error in your numerical solution behave under grid
+   refinement? Can you explain what you see?
+{{< /question >}}

```

Comment:
How are they to solve the problem in this step? I think we want an implementation using the petsc4py stuff and then solving with multigrid? Does it therefore make sense to implement the parts of `mgsolver` that are missing _first_ (getting the 1D test passing), and then get them to implement the beam and check the convergence, using the mgsolver?